### ISSUE TITLE
Tutorial examples refer to things in _dev packages

### ISSUE THREAD

**Issue by IwanVosloo** [reahl/reahl#3]
Things in _dev packages are not supposed to be installed, hence tutorial examples cannot refer to them - they break if they do.

Seems like the culprit is run fixtures - and the suggested solution is to move such run fixtures to reahl.dev or reahl.webdev instead.

---

Imported from Launchpad using lp2gh.
- date created: 2014-02-27T13:12:30Z
- owner: iv
- assignee: craig-reahl
- the launchpad url was https://bugs.launchpad.net/bugs/1285664

**Comment by IwanVosloo** [comment]
(by craig-reahl)
Moved ConfiguredFixture to reahl-dev and left behind a ContextFixture for use by reahl-component
Merged CleanDatabase and Configured Fixtures

Technically this is one more run fixture that now resides in a  *_dev package:
reahl.component_dev.fixtures:ContextFixture

**Comment by IwanVosloo** [comment]
(by craig-reahl)
Given the changes made, it seems there is still an issue when running the example tests regarding the packaging of _dev that are not installed. If I try to run the tutorial.parameterised2 example I get yet another fixture issue, this time however not run_fixture related though:

# 

## ERROR: Failure: ImportError (No module named sqlalchemysupport_dev.fixtures)

Traceback (most recent call last):
  ...
  File "/tmp/parameterised2/parameterised2_dev/parameterised2tests3.py", line 3, in <module>
    from reahl.web_dev.fixtures import WebFixture
  File "/tmp/env2/local/lib/python2.7/site-packages/reahl/web_dev/fixtures.py", line 33, in <module>
    from reahl.domain_dev.fixtures import PartyModelZooMixin
  File "/tmp/env2/local/lib/python2.7/site-packages/reahl/domain_dev/fixtures.py", line 26, in <module>
    from reahl.sqlalchemysupport_dev.fixtures import SqlAlchemyTestMixin 
ImportError: No module named sqlalchemysupport_dev.fixtures

I think we need to discuss the next part of resolving this issue of running the examples without the *_dev packages. Perhaps we need to think about setting up a development environment for other OS distributions(not just Ubuntu)?

**Comment by IwanVosloo** [comment]
(by iv)
Regarding #1, the idea with ContextFixture was that nothing else (except stuff in reahl.component_dev) would use it.

**Comment by IwanVosloo** [comment]
(by iv)
Regarding comment #2: this is a tuff one. We did not think the _dev packages through properly. The initial idea was that stuff in reahl.<package>_dev packages are the tests for reahl.<package>. That means that it will be available to you when you have the source of reahl-<package> checked out, but would not be installed when you install the egg.

The knock-on effect of that is that things in reahl.<package>_dev should never be re-used from other packages.

Trouble is now that Fixtures are meant to be reusable, and we started creating Mixins for composing Fixtures with. What makes matters worse is that something like reahl-web's _dev depends (for tests) on a specific implementation - like what is provided with reahl-domain, as per your stack trace.

We need to take a good look at what in _dev is meant to be reusable and what not. 
Pending the outcome of that though we may want to be able to install (at least some of the) things in _dev.

I see the following options:
 1) create a separate egg with the reusable _dev stuff in it - akin to <package>-dev debian packages, for example that traditionally would ship C header files. There are 2 options for doing this: 
 1a) create 2 setup.py files from one .reahlproject... 
 1b) split the egg source directories entirely into two separate eggs, with file structure as we had it, and each one its own .reahlproject. The latter is probably the simpler solution, but may have issues regarding possible circular dependencies: <package>.egg may never be dependent on <package>_dev.egg; it may be dependent on <another>_dev for testing purposes only.

 2) is to create two special packages underneath reahl/<package> in the source tree: call one reahl/<package)_tests (say), and the other reahl/<package>_dev. Don't install reahl/<package>_tests, but DO install reahl/<package>_dev. The latter then containing the reusable stuff for testing purposes.

 3) is to just install everything currently in reahl/<package>_dev

For purposes of your issue, I'd first just do (3) and see that that solves the issue. That will prove / disprove the solution - then we can decide how much effort to put into shipping things in different packages etc.
